### PR TITLE
feat: using pipeline registry and tests for transformers

### PR DESCRIPTION
**Pull request by aarnphm**
Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

- minor type ignore stubs
- add support to use `PIPELINE_REGISTRY` and backward compatible to older transformers versions.
- add validators for pipeline task type.
- add more related library version to framework context.

### CODE REVIEW

File: bentoml/_internal/frameworks/transformers.py [hunk 1]
```diff
@@ -402,7 +451,7 @@ class TransformersRunnable(bentoml.Runnable):
         def __init__(self):
             super().__init__()
 
-            available_gpus: str = os.getenv("CUDA_VISIBLE_DEVICES")
+            available_gpus: str = os.getenv("CUDA_VISIBLE_DEVICES", "")
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/transformers.py:454]: This is to fix getenv warning from pyright

File: bentoml/_internal/frameworks/transformers.py [hunk 2]
```diff
@@ -305,39 +354,42 @@ def save_model(
...
+            impl: type = task_definition["impl"]
+            if type(pipeline) != impl:
                 raise BentoMLException(
-                    f"Argument `task_definition` '{task_definition}' does not match pipeline task "
-                    "definition '{SUPPORTED_TASKS[task_name]}'."
+                    f"Argument `pipeline` is not an instance of {impl}. It is an instance of {type(pipeline)}."
                 )
-        else:
-            SUPPORTED_TASKS[task_name] = task_definition
+            if task_name in TASK_ALIASES:
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/transformers.py:380]: if task is an alias, then we should throw an error warning users not to override them.

File: bentoml/_internal/frameworks/transformers.py [hunk 3]
```diff
@@ -70,42 +75,47 @@ def _check_flax_supported() -> None:  # pragma: no cover
...
+    import transformers.utils.import_utils as import_utils
+else:
+    _exc_msg = "transformers is required in order to use module 'bentoml.transformers'. Install transformers with 'pip install transformers'."
+    transformers = LazyLoader(
+        "transformers", globals(), "transformers", exc_msg=_exc_msg
+    )
+    import_utils = LazyLoader(
+        "import_utils", globals(), "transformers.utils.import_utils"
     )
+    _check_flax_supported()
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/transformers.py:89]: provide lazy loading.

Comment (parano) [reply bentoml/_internal/frameworks/transformers.py:89]: no need for lazy loading under type checking

File: bentoml/_internal/frameworks/transformers.py [hunk 4]
```diff
@@ -70,42 +75,47 @@ def _check_flax_supported() -> None:  # pragma: no cover
...
-    tf: t.Tuple[str] = attr.field(
-        validator=[
-            attr.validators.deep_iterable(
-                member_validator=attr.validators.instance_of(str)
-            )
-        ],  # type: ignore
-        factory=(tuple),
+    task: str = attr.field(validator=attr.validators.instance_of(str))
+
+    tf: t.Tuple[str] = attr.field(  # type: ignore (unfinished attr type)
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/transformers.py:98]:
```suggestion
    tf: t.Tuple[str] = attr.field(
```

File: bentoml/_internal/frameworks/transformers.py [hunk 5]
```diff
@@ -70,42 +75,47 @@ def _check_flax_supported() -> None:  # pragma: no cover
...
+        factory=tuple,
     )
-    pt: t.Tuple[str] = attr.field(
-        validator=[
-            attr.validators.deep_iterable(
-                member_validator=attr.validators.instance_of(str)
-            )
-        ],  # type: ignore
-        factory=(tuple),
+    pt: t.Tuple[str] = attr.field(  # type: ignore (unfinished attr type)
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/transformers.py:104]:
```suggestion
    pt: t.Tuple[str] = attr.field(
```

File: bentoml/_internal/frameworks/transformers.py [hunk 6]
```diff
@@ -118,6 +128,24 @@ def get(tag_like: str | Tag) -> Model:
     return model
 
 
+def _get_auto_class(
+    val: str | Iterable[str],
+) -> t.Generator[ext.BaseAutoModelClass, None, None]:
+    if isinstance(val, str):
+        if not hasattr(transformers, val):
+            raise BentoMLException(
+                f"Given {val} is neither exists nor valid transformers.AutoModel"
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/transformers.py:137]:
```suggestion
                f"Given {val} is neither exists nor a valid Transformers AutoModel. For more information, please see https://huggingface.co/docs/transformers/model_doc/auto"
```

File: bentoml/_internal/frameworks/transformers.py [hunk 7]
```diff
@@ -290,10 +319,30 @@ def save_model(
...
+    library_versions: dict[str, str] = {}
+    if _check_flax_supported():
+        jax_versions = _check_flax_supported()
+        if jax_versions is not None and isinstance(jax_versions, tuple):
+            library_versions["jax"] = jax_versions[0]
+            library_versions["flax"] = jax_versions[1]
+    if import_utils.is_torch_available():
+        library_versions["torch"] = get_pkg_version("torch")
+    if import_utils.is_tf_available():
+        library_versions["tensorflow"] = get_pkg_version("tensorflow")
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/transformers.py:338]: provides additional library versioning.

**Comment by aarnphm** [comment]
tests will be added later once PIPELINE_REGISTRY is published in newer version of transformers.

File: bentoml/_internal/frameworks/transformers.py (hunk 3 above)
Comment (aarnphm) [reply bentoml/_internal/frameworks/transformers.py]: solved

File: bentoml/_internal/frameworks/transformers.py [hunk 8]
```diff
@@ -71,39 +78,65 @@ def _check_flax_supported() -> None:  # pragma: no cover
...
-            attr.validators.deep_iterable(
-                member_validator=attr.validators.instance_of(str)
+    default: t.Optional[t.Dict[str, t.Any]] = attr.field(
+        factory=dict, converter=task_default_converter
+    )
+    type: t.Optional[str] = attr.field(
+        validator=attr.validators.optional(
+            attr.validators.and_(
+                attr.validators.instance_of(str),
+                attr.validators.in_(("audio", "image", "multimodal", "text")),
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply bentoml/_internal/frameworks/transformers.py:129]: It is the only pipeline type that is allowed on hugging face anyway. The custom pipeline is supposed to be used for pre and post processing for pipelines. They should still resolve to the format that huggingface define, in which the type has to be the 4 type for now.

File: bentoml/_internal/frameworks/transformers.py [hunk 9]
```diff
@@ -116,6 +149,22 @@ def get(tag_like: str | Tag) -> Model:
...
+def _get_auto_class(val: str | Iterable[str]) -> AutoClassGenerator:
+    if isinstance(val, str):
+        if not hasattr(transformers, val):
+            raise BentoMLException(
+                f"Given {val} is neither exists nor a valid Transformers auto class. For more information, please see https://huggingface.co/docs/transformers/model_doc/auto"
+            )
+        yield getattr(transformers, val)
+    elif isinstance(val, Iterable):
+        for v in val:
+            yield from _get_auto_class(v)
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply bentoml/_internal/frameworks/transformers.py:161]: for iterable we should yield the model class itself from transformers.
I guess we can either return a tuple or create a generator here. It is more about preference.

**Comment by aarnphm** [comment]
Closing this for now. Wait until next releases for more mature custom pipeline API upstream.

**Pull request closed without merging**